sig-network/api-extensions-best-practices.md

```diff
@@ -0,0 +1,145 @@
+enhancement proposal process specific to the project.
+
+You can see examples of this with some of the current sub-projects:
+
+* [Gateway Enhancement Proposals (GEPs)][geps]
+* [Network Policy Enhancement Proposals (NPEPs)][npeps]
+
+Key characteristics of this pattern are:
+
+* an easy to find directory at the root of repositories, such as `geps/`, with
+  references provided from the `README.md`.
+* a readme for the enhancement process (we recommend using GEP as a template,
+  and we recommend a motivation focused proposal process (see more below)).
+* a project board which organizes these GEPs and helps provide visibility into
+  the project's priorities and roadmap
+* trackable issues and PRs for every proposal that go through phases (e.g.
+  (`provisional`, `implementable`, `experimental`, e.t.c.)
+
+We generally recommend this pattern unless there's a strong reason not to.
+
+> **Note**: When starting new projects, we advise having enhancement proposals
+> preceed any code whatsoever, for posterity and to help ensure inclusion.
+
+> **Note**: Taking this approach doesn't mean complete autonomy. Any APIs
+> (extension, or otherwise) **MUST** go through the API review process with the
+> SIG Network leads for any content considered stable/standard/GA. Experimental
+> features, alpha, beta, e.t.c. can be developed without that additional
+> oversight until they are promoted.
+
+[KEP]:https://github.com/kubernetes/enhancements/blob/master/keps/README.md
+[geps]:https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/overview.md
+[npeps]:https://github.com/kubernetes-sigs/network-policy-api/tree/main/npeps
+
+#### Motivation Focused Enhancement Proposals
+
+It can be very easy to bring a new proposal to the table, and start with a huge
+GEP and maybe a ton of upfront code. In our experience however having too much
+up front, and in particular coming right to the table with an implementation
+can cause a proposal to get gridlocked very quickly.
+
+**We strongly recommend a "What, why and who before how" approach**. The first
+PR which adds an enhancement proposal should establish:
+
+* **What?** - What do we want to do, at a high level? A summary and/or specific
+  high-level goals to achieve.
+* **Why?** - The motivation, the user stories, the reason we want to do this.
+* **Who?** - Every proposal stands a better chance of succeeding if there are
+  multiple aligned parties interested in it, and collaborating on it upfront.
+
+Without any indication of _how_ we accomplish it. This, plus a process of steps
+for graduation (e.g. `provisional` -> `implementable` -> `experimental` ->
+`standard`) allows the community to work more iteratively, and work on alignment
+with the principles before spending any time in the implementation details.
+
+### Developing CRDs
+
+TODO
+
+### Delivering CRDs
+
+TODO
+
+### Conformance Levels
+
+TODO - from Gateway API conformance levels
```

Comment:
So while "experimental" vs "standard" is a useful distinction, I think that "core" vs "extended" is generally an anti-pattern. Having a standard API is only useful if... you have a standard API.

I think "core" vs "extended" can only be justified if

1. It is assumed that most implementations of the API will rely on some external "backend" that actually does the bulk of the work (eg, Gateway being built on top of nginx or GCP load balancers; or AdminNetworkPolicy being built on top of OVN ACLs or AWS security groups), _AND_
2. There is a feature that everyone agrees really needs to be in the API in order for the API to be viable, _AND_
3. Some "backends" can't implement that desired feature, and it is not reasonable to say "well then I guess you can't use that backend to implement our API" because either
   1. the choice of backend is inherently constrained in some environments (eg, you can't tell people on AWS to just use the GCP-based implementation of Gateway instead), _OR_
   2. the choice of backend was made well before the new feature was added (ie, a newly-proposed feature can't be implemented by some existing implementations of the API, and you don't want to retroactively force them into non-compliance).

But this should be thought of more as a regrettable one-off sort of thing than a normal feature of the API development process.

In particular, "core" vs "extended" is _not_ appropriate for cases like "some people don't want to implement this feature because they don't think it's worth the effort to do so".

Reply:
(Note: this comment is more inspired by discussions I've seen within NPAPI WG than by Gateway, which I'm less familiar with. The only "extended" Gateway feature I know off the top of my head is regex matching, which I believe meets the criteria above.)

Reply:
"antipattern" is an interesting word choice here because it means a common pattern that is ineffective, or has notable downsides. On one hand, I don't think this pattern is all that common (in Gateway API we invented it (or at least, we weren't aware of others doing it at the time)). On the other hand, I would not characterize our experience of using this pattern over the years to be particularly ineffective. It does have the known downside regarding users being able to know when their implementation does (or does not) support a feature, for which we have been [considering solutions](https://github.com/kubernetes-sigs/gateway-api/issues/3164) which is definitely notable.

> In particular, "core" vs "extended" is not appropriate for cases like "some people don't want to implement this feature because they don't think it's worth the effort to do so".

In fact, the motivation for "extended" was because we knew there were features some implementations would not be able to support (potentially _ever_), and we made an intentional trade off to achieve greater API flexibility. Having these features as "extended" has historically provided both motivation AND time relief for implementations to be conformant without having to stop the world to add a major feature right away, though the user experience could be better.

Overall though I can see how this could be contentious, particularly given the fact that the solutions for the known issues with it in Gateway API are not fully complete. I've opted to simply remove this section for now.

> **Note**: As a note in Gateway API a feature is only "extended" when it has conformance test coverage and at least three implementations support it.

Reply:
I agree there are downsides and UX challenges with "extended" functionality that haven't fully been solved, but I actually think it has been a very positive pattern for facilitating many implementations to engage with the Gateway API project who may not have otherwise, and I would like to see this pattern included in guidance.

To clarify, "extended" functionality in Gateway API [_is covered by conformance tests_](https://gateway-api.sigs.k8s.io/concepts/conformance/#2-support-levels) - it's optional for implementations, but if implemented must behave in the expected manner. This pattern is used extensively for throughout Gateway API, including:
- [HTTPRouteRule](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPRouteRule) names, timeouts, retries and session persistence configuration
- Many [HTTPRouteFilter](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteFilter) types
as well as smaller details like:
- Specifying a `port` field on [ParentReference](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.ParentReference)
- Service `appProtocol` support in [HTTPBackendRef](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPBackendRef)

I would contend that having common APIs supported by _some_ vendors is a better alternative than either a proliferation of bespoke vendor-specific APIs or an expansive required core which only a limited number of implementations may be able to support, especially with the intent of promoting sufficiently-adopted "extended" functionality into "core" when applicable.

I've actually wanted to extend the use of this pattern to other subprojects where I believe it could alleviate contention around desired (but arguably not essential) functionality that has been a hindrance to adoption by some projects - two specific examples in AdminNetworkPolicy would be a ["passthrough" audit/dry-run mode](https://github.com/kubernetes-sigs/network-policy-api/issues/230) (currently not possible even though it has been a [requested feature](https://github.com/cilium/cilium/issues/9580) which the project could eventually choose to implement), and [priority levels](https://github.com/cilium/cilium/issues/23380#issuecomment-1423328510), which have been barriers to Cilium's potential adoption of this spec.

This is notably distinct from an "implementation-specifc" pattern (as mentioned for RegEx filters), which are intentionally excluded from conformance due to irreconciallable dataplane implementations.

Reply:
> I don't think this pattern is all that common

> I've opted to simply remove this section for now.

I brought this up specifically because ANP copied it from Gateway, and I very strongly believe that they should not have, and most other WGs should not either. (So I think we should talk about it in this document, and you should un-mark-this-conversation-resolved.)

> I actually think it has been a very positive pattern for facilitating many implementations to engage with the Gateway API project

The goal of SIG Network WGs is not to get implementations to engage with SIG Network, it is to provide useful networking features for end users of Kubernetes. Obviously we like having engaged implementers, and obviously a vibrant Kubernetes ecosystem is good for end users, but _just_ having more projects engage with Gateway API is not an inherently good thing if it results in an API that does not help users get stuff done in their clusters.

> I would contend that having common APIs supported by some vendors is a better alternative than either a proliferation of bespoke vendor-specific APIs

Why? If you don't have a standard, then users can't use the API in a vendor-agnostic manner anyway, so what does it matter if they have to use vendor-specific APIs?

If most vendors don't implement the feature, then you're basically going to have to design your cluster around a single-vendor solution, so it doesn't matter much if you have a standard API or a vendor-specific API.

If most vendors _do_ implement the feature, then people will just think of it as being a standard feature, and then the vendors that don't implement it essentially just become buggy in users' eyes.

IMO, if a feature is not available to all users of the API, then it should _look_ like it's not available to all users of the API.

I know that we fail at this with basically all of our multiply-implemented APIs, but as I said above, I think this should be something we regret, not something we pursue.

> two specific examples in AdminNetworkPolicy which have been barriers to Cilium's potential adoption of this spec would be a "passthrough" audit/dry-run mode

Right, and I feel like that feature should not be part of ANP

> and priority levels

Whereas for that, Cilium either needs to implement priority levels, or just not implement ANP. Priority levels are not a separable part of the semantics of ANP. If you get rid of priority then you also have to get rid of mixed Allow/Deny, and then you're basically back at NPv1 again and not solving any of the ANP user stories.

Reply:
@danwinship Thanks for clarifying your concerns regarding the "extended" functionality pattern, agreed we should at least document the status quo and current lack of consensus on a recommended path forward.

Reply:
> So I think we should talk about it in this document, and you should un-mark-this-conversation-resolved

I resolved this and removed the section because I immediately realized that this is a larger conversation that I wouldn't want to hold up any other potential value in this PR. I think this is worth it's own dedicated discussion so that it can be analyzed and debated with great focus.

Reply:
> The goal of SIG Network WGs is not to get implementations to engage with SIG Network, it is to provide useful networking features for end users of Kubernetes.

❤️

Reply:
So on the specific point of:

> The goal of SIG Network WGs is not to get implementations to engage with SIG Network, it is to provide useful networking features for end users of Kubernetes.

There's nuance here, I'm not sure I'm clear on the point of this distinction. Also I think "sub-project" and not WG may have been meant. In any case, I see these two things as intertwined.

Gateway API was in a unique (at the time) position of needing to find success by means _other_ than by virtue of being available by default. Given that position I don't see how it could have ever even made it into the hands of users without getting implementations to engage with SIG Network.

Gateway API engaged with implementations in order to collaboratively build the networking features users can now take for granted from it in Kubernetes. These implementations (through their contributors) are now a _part of SIG Network_.

In a future where we say the bulk of new and experimental networking features must take a CRD-based approach for development and distribution I think it will mostly mean by definition that those projects will need implementations to engage with SIG Network in order to succeed, in service of the greater goal of providing useful networking features for end users of Kubernetes.

For the purposes of what this document is getting at: perhaps another way to phrase this "The goal of SIG Network Sub-Projects is to provide useful networking features for end users of Kubernetes. A requirement to meet that goal for CRD-based implementations is often to get implementations to engage with SIG Network".

Reply:
So on the specific point of:

> The goal of SIG Network WGs is not to get implementations to engage with SIG Network, it is to provide useful networking features for end users of Kubernetes.

There's nuance here, I'm not sure I'm clear on the point of this distinction. Also I think "sub-project" and not WG may have been meant. In any case, I see these two things as intertwined.

Gateway API was in a unique (at the time) position of needing to find success by means _other_ than by virtue of being available by default. Given that position I don't see how it could have ever even made it into the hands of users without getting implementations to engage with SIG Network.

Gateway API engaged with implementations in order to collaboratively build the networking features users can now take for granted from it in Kubernetes. These implementations (through their contributors) are now a _part of SIG Network_.

In a future where we say the bulk of new and experimental networking features must take a CRD-based approach for development and distribution I think it will mostly mean by definition that those projects will need implementations to engage with SIG Network in order to succeed, in service of the greater goal of providing useful networking features for end users of Kubernetes.

For the purposes of what this document is getting at I would say "The goal of SIG Network Sub-Projects is to provide useful networking features for end users of Kubernetes. A requirement to meet that goal for CRD-based implementations is often to get implementations to engage with SIG Network". And to the point that I think Dan was getting at, "that engagement has to be careful to stay aligned with what's best for the user" and it's unclear whether conformance levels are in best service of the user.

The above situation definitely relates well to a "best practices" guide, as we would want to provide guidance on both:

1. How to remain focused on the user, and avoid the temptations to do things that might actually serve the developers more.
2. Whether or not we would advise providing "conformance levels" in any new APIs, as it pertains to `#1`.

I don't think we have a definitive answer on `#2` (and there might be a lot of discussion to have there), so while that is in doubt I'm fine with documenting it here as an "anti-pattern" for now and warning new projects against it with the note that there is ongoing discussion and efforts to level it up (and then create that discussion in the [Gateway API discussion boards](https://github.com/kubernetes-sigs/gateway-api/discussions)).

Is that sufficient from your perspective to consider this comment thread resolved @danwinship?

Reply:
> warning new projects against it

That is mostly what I want; NP API basically just copied everything that Gateway did to get started, and I don't want people copying core-vs-extended without thinking about whether it actually makes sense for them.

Reply:
1bae0c13f12bdb9ca142a41c91678e938e0b8b32 adds a section to this effect, LMKWYT.